sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1372,6 +1377,96 @@ object PushPredicateThroughJoin extends Rule[LogicalPlan] with PredicateHelper {
   }
 }
 
+/**
+ * Rewriting join condition to conjunctive normal form expression so that we can push
+ * more predicate.
+ */
+object PushPredicateThroughJoinByCNF extends Rule[LogicalPlan] with PredicateHelper {
+  /**
+   * Rewrite pattern:
+   * 1. (a && b) || (c && d) --> (a || c) && (a || d) && (b || c) && (b && d)
```

Comment:
1 .  (a && b) || (c && d) --> (a || c) && (a || d) && (b || c) && (b || d) ？

Reply:
Thank you. I will update it in the next commit.